src/crimson/os/seastore/journal/circular_bounded_journal.cc

```diff
@@ -0,0 +1,569 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include <boost/iterator/counting_iterator.hpp>
+
+#include "crimson/common/errorator-loop.h"
+#include "include/intarith.h"
+#include "crimson/os/seastore/journal/circular_bounded_journal.h"
+#include "crimson/os/seastore/logging.h"
+
+SET_SUBSYS(seastore_journal);
+
+namespace crimson::os::seastore::journal {
+
+std::ostream &operator<<(std::ostream &out,
+    const CircularBoundedJournal::cbj_header_t &header)
+{
+  return out << "cbj_header_t(magin=" << header.magic
+	     << ", uuid=" << header.uuid
+	     << ", block_size=" << header.block_size
+	     << ", size=" << header.size
+	     << ", used_size=" << header.used_size
+	     << ", error=" << header.error
+	     << ", start_offset=" << header.start_offset
+	     << ", applied_to="<< header.applied_to
+	     << ", last_committed_record_base="<< header.last_committed_record_base
+	     << ", written_to=" << header.written_to
+	     << ", flsg=" << header.flag
+	     << ", csum_type=" << header.csum_type
+	     << ", csum=" << header.csum
+	     << ", start=" << header.start
+	     << ", end=" << header.end
+             << ")";
+}
+
+
+CircularBoundedJournal::CircularBoundedJournal(NVMeBlockDevice* device,
+    const std::string path)
+  : device(device), path(path) {}
+
+CircularBoundedJournal::mkfs_ret
+CircularBoundedJournal::mkfs(mkfs_config_t& config)
+{
+  LOG_PREFIX(CircularBoundedJournal::mkfs);
+  return _open_device(path
+  ).safe_then([this, &config, FNAME]() -> mkfs_ret {
+    rbm_abs_addr start_addr = convert_paddr_to_abs_addr(
+      config.start);
+    assert(config.block_size == device->get_block_size());
+    ceph::bufferlist bl;
+    CircularBoundedJournal::cbj_header_t head;
+    head.magic = CBJOURNAL_MAGIC;
+    head.uuid = uuid_d(); // TODO
+    head.block_size = config.block_size;
+    rbm_abs_addr end_addr = convert_paddr_to_abs_addr(
+      config.end);
+    head.size = end_addr - start_addr
+      - device->get_block_size();
+    head.used_size = 0;
+    head.error = 0;
+    head.start_offset = device->get_block_size();
+    head.last_committed_record_base = 0;
+    head.written_to = head.start_offset;
+    head.applied_to = 0;
+    head.flag = 0;
+    head.csum_type = 0;
+    head.csum = 0;
+    head.cur_segment_seq = 0;
+    head.start = start_addr;
+    head.end = end_addr;
+    head.device_id = config.device_id;
```

Comment:
Most of this should be part of the default constructor for cbj_header_t.  Seems like we can only update the things that are actually non-default.